src/catchup/DownloadApplyTxsWork.cpp

```diff
@@ -43,44 +43,40 @@ DownloadApplyTxsWork::yieldMoreWork()
     {
         throw std::runtime_error("Work has no more children to iterate over!");
     }
+    std::vector<FileType> fileTypesToDownload{
+        FileType::HISTORY_FILE_TYPE_TRANSACTIONS};
+    std::vector<std::shared_ptr<BasicWork>> downloadSeq;
+    std::vector<FileTransferInfo> filesToTransfer;
+    if (mApp.getConfig().CATCHUP_SKIP_KNOWN_RESULTS)
+    {
+        fileTypesToDownload.emplace_back(FileType::HISTORY_FILE_TYPE_RESULTS);
+    }
+    for (auto const& fileType : fileTypesToDownload)
+    {
+        CLOG_INFO(History,
+                  "Downloading, unzipping and applying {} for checkpoint {}",
+                  typeString(fileType), mCheckpointToQueue);
+        FileTransferInfo ft(mDownloadDir, fileType, mCheckpointToQueue);
+        filesToTransfer.emplace_back(ft);
+        downloadSeq.emplace_back(
+            std::make_shared<GetAndUnzipRemoteFileWork>(mApp, ft, mArchive));
+    }
 
-    CLOG_INFO(History,
-              "Downloading, unzipping and applying {} for checkpoint {}",
-              typeString(FileType::HISTORY_FILE_TYPE_TRANSACTIONS),
-              mCheckpointToQueue);
-    FileTransferInfo ft(mDownloadDir, FileType::HISTORY_FILE_TYPE_TRANSACTIONS,
-                        mCheckpointToQueue);
-    auto getAndUnzip =
-        std::make_shared<GetAndUnzipRemoteFileWork>(mApp, ft, mArchive);
+    OnFailureCallback cb = [archive = mArchive, filesToTransfer]() {
```

Comment:
Please keep the implementation of OnFailureCallback the way it was before: mArchive can be null, in which case `GetAndUnzipRemoteFileWork` will pick a random archive. With the current change, core would crash de-referencing a null pointer.